### ISSUE TITLE
Add support for simple monolithic & vanilla docker setup of test environment

### ISSUE THREAD

**Issue by benefacto** [polywrap/monorepo#95]
**TL:DR**:
Deliver 1 & 3 in [this diagram](https://github.com/Web3-API/monorepo/issues/95#issuecomment-772089653)

**Goals:**
- Easily & consistently setup web3api test environments, enabling automated regression tests & code coverage

**Review current system:**
- [x] Identify where & how code is currently built and current test environment is setup:
1. Environment is spun up via the Web3API CLI: `npx w3 test-env up` but that is currently just a wrapper for [a static Docker compose file](https://github.com/Web3-API/prototype/blob/master/packages/client-test-env/docker-compose.yml) ([the web3api.yaml file](https://github.com/Web3-API/prototype/blob/master/demos/simple-storage/protocol/web3api.yaml) is not used for this)
2. Contracts are deployed via `node ./deploy-contracts.js`: 
- Contracts are deployed [a hardcoded local Ethereum provider](https://github.com/Web3-API/prototype/blob/5d61039199873089f6ca8e02c4e6146362447e4e/demos/simple-storage/protocol/deploy-contracts.js#L9) via `@web3api/client-js` ([the same one](https://github.com/Web3-API/prototype/blob/034b8c2decb646e677d252da5c273f18934e016d/packages/client-test-env/.env#L2) setup in the previous step)
- The subgraph manifest file is updated with the [deployed address of the contract](https://github.com/Web3-API/prototype/blob/5d61039199873089f6ca8e02c4e6146362447e4e/demos/simple-storage/protocol/deploy-contracts.js#L22)
- The recipe constants file is updated with the [deployed address of the contract](https://github.com/Web3-API/prototype/blob/5d61039199873089f6ca8e02c4e6146362447e4e/demos/simple-storage/protocol/deploy-contracts.js#L32)

**Refactor Docker Compose setup:**
~~Break out [current single compose file](https://github.com/Web3-API/prototype/blob/prealpha-dev/packages/test-env/docker-compose.yml) into [multiple compose files](https://docs.docker.com/compose/extends/#multiple-compose-files) for each service with the dev server service as the base one, updating [the test-env compose commands](https://github.com/Web3-API/prototype/blob/daed9cf2c22053a838b85c6a14857a3bf7167803/packages/test-env/index.js#L29) to reference these files~~
-~~Add optional boolean parameters to the test-env commands that correspond to each of these files that default to true~~
- [x] Revert CLI changes
- [x] [Add](https://docs.docker.com/docker-hub/repos/#pushing-a-docker-container-image-to-docker-hub) Web3API docker images [here](https://hub.docker.com/repository/docker/benefacto/web3-api_dev-server)
- [x] Change CLI to use docker files in current directory by default
- [x] Add passing in of path to docker files to CLI
- [x] Add docker-compose files to example project
- [x] Resolve linting errors
- [x] Add CI parameter that uses test-env docker file, add this to all CI calls
- [x] Replace directory parameter with second parameter, default to the current directory if not present

**Test Changes:**
- [x] Identify where & how tests are currently run
The Web3API project is tested via the Web3API CLI: npx w3 query ./recipes/e2e.json --test-ens 
- [x] Add test CLI parameters, something like: `w3 test-recipes` with parameters for endpoints with default values

**Rework Docker Compose setup:**
- [x] Use web3api.env.yml file as a singular docker-compose file in demo project
- [x] Set web3api.env.yml as default file to look for in test commands: `w3 test up` & `w3 test down`
- [x] Fix failing CI test error
- [x] Create ipfs docker-compose file in test-env package
- [x] Create ganache-cli docker-compose file test-env package
- [x] Set additional arguments as modules in test commands, such as: `w3 test up ipfs ethereum` & `w3 test down ipfs ethereum`
- [x] Add validation to passed modules
- [x] Create graph-node docker-compose file test-env package
- [x] Use dev-server docker service by default, removing reference to image

**Comment by dOrgJelli** [comment]
Some high level goals of this:
- Easily & consistently build web3api projects from their source files, enabling source verification
- Easily & consistently setup web3api test environments, enabling automated regression tests & code coverage

**Comment by benefacto** [comment]
It would also be nice if the test environment was shutdown after a period of time or had some sort of auto shutdown parameter. I've noticed on my machine that it seems to be a bit of a bandwidth hog.

**Comment by dOrgJelli** [comment]
There is a `w3 test-env down` command

**Comment by benefacto** [comment]
**Core Sync Feedback**
- ~~Snapshot build environment (e,g, [`docker commit`](https://docs.docker.com/engine/reference/commandline/commit/)) before compiling to ensure reproducibility~~ split to #121 
- ~~Should have separate configuration files for build and test setup, also for environments~~ split to #124 
- ~~Files could be docker-compose files that extend from our default docker-compose files~~ split to #124

**Comment by dOrgJelli** [comment]
Just had a great call with Ben where we refined the developer experience. We're imagining 3 scenarios:
1. Super simple hyper abstracted setup, where a user a selecting pre-configured services from a monolithic docker-compose file.
2. Compose multiple docker-compose "fragments" into a single file that will combine and instantiate them all.
3. Use the docker-compose format exactly, and just rename the file to web3api.env.yaml.

We're thinking that 1 & 3 can get done in the same PR, and we'll add 2 capabilities in the future. Please add any additional feedback or thoughts you have.

![image](https://user-images.githubusercontent.com/5522128/106676565-c88d8500-657c-11eb-95b8-33bc9e9d1b33.png)

**Comment by dOrgJelli** [comment]
Here's an article where they discuss modular docker-compose files: https://blog.software-craftsmen.at/blog/2017/docker-compose-best-practices-1.html

**Comment by krisbitney** [comment]
It would be great to add support for additional docker images and to be able to change ports if necessary to avoid conflicts. For example, adding this text to the docker compose file produces a ganache mainnet fork that is helpful for testing:

```
  mainnet-fork:
    image: 'trufflesuite/ganache-cli'
    ports:
      - '8546:8545'
    command: -l 8000000 --deterministic --hostname=0.0.0.0 --chainId 1 --fork https://mainnet.infura.io/v3/{project-id}
```

**Comment by cbrzn** [comment]
I guess this can be closed since infra has been developed? @dOrgJelli @namesty
